src/topology/algebra/weak_dual_topology.lean

```diff
@@ -0,0 +1,191 @@
+  it is sufficient that its compositions with evaluations at all points `z : E` are continuous
+* `tendsto_iff_forall_eval_tendsto` is a characterization of convergence in `weak_dual 𝕜 E` in
+  terms of convergence of the evaluations at all points `z : E`
+
+## Notations
+
+No new notation is introduced.
+
+## Implementation notes
+
+The weak-* topology is defined as the induced topology under the mapping that associates to a dual
+element `x'` the functional `E → 𝕜`, when the space `E → 𝕜` of functionals is equipped with the
+topology of pointwise convergence (product topology).
+
+The general definition assumes that `𝕜` is a topological semiring in the sense of the typeclasses
+ `topological_space 𝕜`, `semiring 𝕜`, `has_continuous_add 𝕜`, `has_continuous_mul 𝕜`,
+and that the space `E` is a topological module over `𝕜` in the sense of the typeclasses
+`topological_space E`, `add_comm_monoid E`, `has_continuous_add E`, `module 𝕜 E`,
+`has_continuous_smul 𝕜 E`.
+
+## References
+
+* https://en.wikipedia.org/wiki/Weak_topology#Weak-*_topology
+
+## Tags
+
+weak-star, weak dual
+
+-/
+
+noncomputable theory
+open filter
+open_locale topological_space
+
+section weak_star_topology
+/-!
+### Weak star topology on duals of topological modules
+In this section, we define the weak-* topology on duals of suitable topological modules `E` over
+suitable topological semirings `𝕜`. The (weak) dual `weak_dual 𝕜 E` consists of continuous linear
+functionals `E →L[𝕜] 𝕜` from `E` to scalars `𝕜`. The weak-* topology is the coarsest topology on
+this dual `weak_dual 𝕜 E := (E →L[𝕜] 𝕜)` w.r.t. which the evaluation maps at all `z : E` are
+continuous.
+
+The weak dual is a module over `𝕜` if the semiring `𝕜` is commutative.
+-/
+
+variables (𝕜 : Type*) [topological_space 𝕜] [semiring 𝕜]
+variables [has_continuous_add 𝕜] [has_continuous_mul 𝕜]
+variables (E : Type*) [topological_space E] [add_comm_monoid E] [module 𝕜 E]
+variables [has_continuous_add E]
+variables [has_continuous_smul 𝕜 E]
+
+/-- The (weak) dual of a topological module `E` over a topological semiring `𝕜` consists of
+continuous linear functionals from `E` to scalars `𝕜`. It is a type synonym with the original
+dual, but will be equipped with a different topology. -/
+@[derive [inhabited, has_coe_to_fun]]
+def weak_dual := E →L[𝕜] 𝕜
+
+instance [has_continuous_add 𝕜] : add_comm_monoid (weak_dual 𝕜 E) := continuous_linear_map.add_comm_monoid
+
+namespace weak_dual
+
+/-- The weak-* topology instance `weak_dual_topology` on the dual of a topological module `E` over
+a topological semiring `𝕜` is defined as the induced topology under the mapping that associates to
+a dual element `x' : weak_dual 𝕜 E` the functional `E → 𝕜`, when the space `E → 𝕜` of functionals
+is equipped with the topology of pointwise convergence (product topology). -/
+instance : topological_space (weak_dual 𝕜 E) :=
+topological_space.induced (λ x' : weak_dual 𝕜 E, λ z : E, x' z) Pi.topological_space
+
+lemma eval_continuous' :
```

Comment:
Shouldn't this be
```suggestion
lemma coe_fn_continuous' :
```

Reply:
@eric-wieser I think `coe_fn_continuous` is ambiguous:  does it mean that the coercion from `weak_dual 𝕜 E` to `E → 𝕜` is continuous (indeed, that's what it's supposed to mean), or that for any `f : weak_dual 𝕜 E`, the function `coe_fn f` of type `E → 𝕜` is continuous (not what it's supposed to mean)?

I agree `eval_continuous'` is also not quite clear, but I think it's better to be unclear by having no obvious meaning than by having two obvious potential meanings.

Reply:
If you write it as the defeq
```lean
@continuous (weak_dual 𝕜 E) (E → 𝕜) _ _ coe_fn
```
that is, fully simplified, then hopefully the name `coe_fn_continuous` is obvious.

Reply:
I don't know which one is most in line with mathlib naming, but I kind of like `coe_fn_continuous` (without a prime then, right?) if the next lemma is `eval continuous`. (My superficial reasons: these names are quite descriptive of the content + avoids primes in names.)

Thank you! I will make the change.